presto-main/src/main/java/com/facebook/presto/operator/MergeOperator.java

```diff
@@ -0,0 +1,502 @@
+    private final MergeSortComparator comparator;
+    private final List<Integer> outputChannels;
+    private final List<Type> outputTypes;
+    private final PageBuilder pageBuilder;
+
+    private final SettableFuture<Void> blockedOnSplits = SettableFuture.create();
+
+    private final Set<URI> locations = new HashSet<>();
+    private final Closer closer = Closer.create();
+    private MergeSources mergeSources;
+
+    private boolean closed;
+
+    public MergeOperator(
+            OperatorContext operatorContext,
+            PlanNodeId sourceId,
+            Supplier<ExchangeClient> exchangeClientSupplier,
+            PagesSerde pagesSerde,
+            MergeSortComparator comparator,
+            List<Integer> outputChannels,
+            List<Type> outputTypes)
+    {
+        this.operatorContext = requireNonNull(operatorContext, "operatorContext is null");
+        this.sourceId = requireNonNull(sourceId, "sourceId is null");
+        this.exchangeClientSupplier = requireNonNull(exchangeClientSupplier, "exchangeClientSupplier is null");
+        this.pagesSerde = requireNonNull(pagesSerde, "pagesSerde is null");
+        this.comparator = requireNonNull(comparator, "comparator is null");
+        this.outputChannels = requireNonNull(outputChannels, "outputChannels is null");
+        this.outputTypes = requireNonNull(outputTypes, "outputTypes is null");
+        this.pageBuilder = new PageBuilder(outputTypes);
+    }
+
+    @Override
+    public PlanNodeId getSourceId()
+    {
+        return sourceId;
+    }
+
+    @Override
+    public Supplier<Optional<UpdatablePageSource>> addSplit(Split split)
+    {
+        requireNonNull(split, "split is null");
+        checkArgument(split.getConnectorId().equals(REMOTE_CONNECTOR_ID), "split is not a remote split");
+
+        URI location = ((RemoteSplit) split.getConnectorSplit()).getLocation();
+        locations.add(location);
+        return Optional::empty;
+    }
+
+    @Override
+    public void noMoreSplits()
+    {
+        ImmutableList.Builder<MergeSource> builder = ImmutableList.builder();
+        for (URI location : locations) {
+            ExchangeClient exchangeClient = closer.register(exchangeClientSupplier.get());
+            exchangeClient.addLocation(location);
+            exchangeClient.noMoreLocations();
+            builder.add(new MergeSource(location, exchangeClient, pagesSerde));
+        }
+
+        mergeSources = new MergeSources(builder.build());
+        blockedOnSplits.set(null);
+    }
+
+    @Override
+    public OperatorContext getOperatorContext()
+    {
+        return operatorContext;
+    }
+
+    @Override
+    public List<Type> getTypes()
+    {
+        return outputTypes;
+    }
+
+    @Override
+    public void finish()
+    {
+        close();
+    }
+
+    @Override
+    public boolean isFinished()
+    {
+        return closed || (mergeSources != null && mergeSources.isFinished() && pageBuilder.isEmpty());
```

Comment:
MergeOperator doesn't accept any pages as an input. In case of this operator `finish()` is used to notify that the operator above doesn't need any input anymore (user canceled the query, the operator above it is limit operator, etc.). What is done in `MergeOperator` is exactly the same what you can see in `ExchangeOperator`.

Even if i removed the `closed` flag check - it will still be the same. Because on finish all the http result fetchers are being closed. So, effectively `closed` flag is just a shortcut.